packages/flutter/lib/src/material/button_style.dart

```diff
@@ -0,0 +1,513 @@
+///         return null; // Use the component's default.
+///       },
+///     ),
+///   ),
+/// )
+///```
+///
+/// In this case the background color for all other button states would fallback
+/// to the ContainedButton’s default values. To unconditionally set the button's
+/// [backgroundColor] for all states one could write:
+///
+/// ```dart
+/// ContainedButton(
+///   style: ButtonStyle(
+///     backgroundColor: MaterialStateProperty.all<Color>(Colors.green),
+///   ),
+/// )
+///```
+///
+/// Configuring a ButtonStyle directly makes it possible to very
+/// precisely control the button’s visual attributes for all states.
+/// This level of control is typically required when a custom
+/// “branded” look and feel is desirable.  However, in many cases it’s
+/// useful to make relatively sweeping changes based on a few initial
+/// parameters with simple values. The button styleFrom() methods
+/// enable such sweeping changes. See for example:
+/// [TextButton.styleFrom], [ContainedButton.styleFrom],
+/// [OutlinedButton.styleFrom].
+///
+/// For example, to override the default text and icon colors for a
+/// [TextButton], as well as its overlay color, with all of the
+/// standard opacity adjustments for the pressed, focused, and
+/// hovered states, one could write:
+///
+/// ```dart
+/// TextButton(
+///   style: TextButton.styleFrom(primary: Colors.green),
+/// )
+///```
+///
+/// To configure all of the application's text buttons in the same
+/// way, specify the overall theme's `textButtonTheme`:
+/// ```dart
+/// MaterialApp(
+///   theme: ThemeData(
+///     textButtonTheme: TextButtonThemeData(
+///       style: TextButton.styleFrom(primary: Colors.green),
+///     ),
+///   ),
+///   home: MyAppHome(),
+/// )
+///```
+/// See also:
+///
+///  * [TextButtonTheme], the theme for [TextButton]s.
+///  * [ContainedButtonTheme], the theme for [ContainedButton]s.
+///  * [OutlinedButtonTheme], the theme for [OutlinedButton]s.
+@immutable
+class ButtonStyle with Diagnosticable {
+  /// Create a [ButtonStyle].
+  const ButtonStyle({
+    this.textStyle,
+    this.backgroundColor,
+    this.foregroundColor,
+    this.overlayColor,
+    this.shadowColor,
+    this.elevation,
+    this.padding,
+    this.minimumSize,
+    this.side,
+    this.shape,
+    this.mouseCursor,
+    this.visualDensity,
+    this.tapTargetSize,
+    this.animationDuration,
+    this.enableFeedback,
+  });
+
+  /// The style for a button's [Text] widget descendants.
+  ///
+  /// The color of the [textStyle] is typically not used directly, the
+  /// [foreground] color is used instead.
+  final MaterialStateProperty<TextStyle> textStyle;
+
+  /// The button's background fill color.
+  final MaterialStateProperty<Color> backgroundColor;
+
+  /// The color for the button's [Text] and [Icon] widget descendants.
+  ///
+  /// This color is typically used instead of the color of the [textStyle]. All
```

Comment:
I qualified a bunch of the API doc in this class with "typically" because it's up to the subclasses and even though the subclasses I've defined are typical, there's no reason to assume that all subclasses will be.